File: guests/js/pickers.ts
// The pickers scene (tools/scenes/pickers.steps; docs/datetime-plan.md).
//     KAYA_SELFTEST=pickers node guests/js/pickers.ts

import * as kaya from "kaya-gui";

const Task = kaya.record({ name: String, due: kaya.CivilDate }, "Task");

const app = new kaya.App();

function day(d: kaya.CivilDate): string {
  return `${String(d.year).padStart(4, "0")}-${String(d.month).padStart(2, "0")}-${String(d.day).padStart(2, "0")}`;
}

function clock(t: kaya.CivilTime): string {
  return `${String(t.hour).padStart(2, "0")}:${String(t.minute).padStart(2, "0")}`;
}

let dateText!: kaya.Signal<string>;
let timeText!: kaya.Signal<string>;
let rowText!: kaya.Signal<string>;
let dateSig!: kaya.Signal<kaya.CivilDate>;
let timeSig!: kaya.Signal<kaya.CivilTime>;

function onDate(picked: kaya.CivilDate): void {
  dateText.set(`date: ${day(picked)}`);
}

function onTime(picked: kaya.CivilTime): void {
  timeText.set(`time: ${clock(picked)}`);
}

function onRowDate(row: kaya.RowHandle<kaya.Fields<typeof Task.schema>>, picked: kaya.CivilDate): void {
  rowText.set(`row ${String(row.key)}: ${day(picked)}`);
}

function onReset(): void {
  dateSig.set({ year: 2026, month: 3, day: 1 });
  timeSig.set({ hour: 9, minute: 0 });
}

app.window({}, () => {
  dateText = kaya.signal("date: none");
  timeText = kaya.signal("time: none");
  rowText = kaya.signal("row: none");
  dateSig = kaya.signal<kaya.CivilDate>({ year: 2026, month: 9, day: 4 });
  timeSig = kaya.signal<kaya.CivilTime>({ hour: 14, minute: 30 });
  const tasks = kaya.collection(Task);
  kaya.column(() => {
    kaya.label({ bind: dateText }); // label#0
    kaya.label({ bind: timeText }); // label#1
    kaya.label({ bind: rowText }); // label#2
    kaya
      .datePicker({ value: dateSig, min: { year: 2026, month: 1, day: 1 }, max: { year: 2026, month: 12, day: 31 }, onChange: onDate })
      .a11yId("when")
      .a11yLabel("Due"); // date_picker#0
    kaya.timePicker({ value: timeSig, step: 15, onChange: onTime }).a11yId("at").a11yLabel("At"); // time_picker#0
    kaya.button("reset", { onClick: onReset }); // button#0
    for (const task of tasks) {
      kaya.label({ bind: task.name });
      kaya.datePicker({ value: task.due, onChange: onRowDate }).a11yId("due");
    }
  });
  tasks.insert("a", Task({ name: "a", due: { year: 2026, month: 10, day: 1 } }));
  tasks.insert("b", Task({ name: "b", due: { year: 2026, month: 11, day: 20 } }));
});

app.run();

File: guests/python/pickers.py
"""The pickers scene (tools/scenes/pickers.steps; docs/datetime-plan.md)."""

import datetime
import sys
from dataclasses import dataclass

import kaya


@dataclass
class Task:
    name: str
    due: datetime.date


app = kaya.App()


def clock(when):
    return f"{when.hour:02d}:{when.minute:02d}"


def on_date(picked):
    date_text.set(f"date: {picked}")


def on_time(picked):
    time_text.set(f"time: {clock(picked)}")


def on_row_date(key, picked):
    row_text.set(f"row {key}: {picked}")


def on_reset():
    date_sig.set(datetime.date(2026, 3, 1))
    time_sig.set(datetime.time(9, 0))


with app.window():
    date_text = kaya.signal("date: none")
    time_text = kaya.signal("time: none")
    row_text = kaya.signal("row: none")
    date_sig = kaya.signal(datetime.date(2026, 9, 4))
    time_sig = kaya.signal(datetime.time(14, 30))
    tasks = kaya.collection(Task)
    with kaya.column():
        kaya.label(bind=date_text)                              # label#0
        kaya.label(bind=time_text)                              # label#1
        kaya.label(bind=row_text)                               # label#2
        kaya.date_picker(                                       # date_picker#0
            value=date_sig,
            min=datetime.date(2026, 1, 1),
            max=datetime.date(2026, 12, 31),
            on_change=on_date,
        ).a11y_id("when").a11y_label("Due")
        kaya.time_picker(                                       # time_picker#0
            value=time_sig, step=15, on_change=on_time,
        ).a11y_id("at").a11y_label("At")
        kaya.button("reset", on_click=on_reset)                 # button#0
        for task in tasks:
            kaya.label(bind=task.name)
            kaya.date_picker(value=task.due,
                             on_change=on_row_date).a11y_id("due")
    tasks.insert("a", Task(name="a", due=datetime.date(2026, 10, 1)))
    tasks.insert("b", Task(name="b", due=datetime.date(2026, 11, 20)))

sys.exit(app.run())
